lib/rhea-promise/index.ts

```diff
@@ -0,0 +1,662 @@
+      connection.removeListener("connection_close", onClose);
+      connection.removeListener("disconnected", onClose);
+      process.nextTick(() => {
+        debug("Resolving the promise with amqp connection.");
+        resolve(connection);
+      });
+    }
+
+    function onClose(context: Context): void {
+      connection.removeListener("connection_open", onOpen);
+      connection.removeListener("connection_close", onClose);
+      connection.removeListener("disconnected", onClose);
+      debug(`Error occurred while establishing amqp connection.`, context.connection.error);
+      reject(context.connection.error);
+    }
+
+    connection.once("connection_open", onOpen);
+    connection.once("connection_close", onClose);
+    connection.once("disconnected", onClose);
+  });
+}
+
+/**
+ * Closes the amqp connection.
+ * @param {Connection} connection The amqp connection that needs to be closed.
+ * @return {Promise<void>} Promise<void>
+ * - **Resolves** the promise when rhea emits the "connection_close" event.
+ * - **Rejects** the promise with an AmqpError when rhea emits the "connection_error" event while trying
+ * to close an amqp connection.
+ */
+export async function closeConnection(connection: any): Promise<void> {
+  return new Promise<void>((resolve, reject) => {
+    if (!connection || (connection && typeof connection !== "object")) {
+      reject(new Error("connection is a required parameter and must be of tyepe 'object'."));
+    }
+
+    if (connection.is_open()) {
+      function onClose(context: Context): void {
+        connection.removeListener("connection_close", onClose);
+        process.nextTick(() => {
+          debug("Resolving the promise as the connection has been successfully closed.");
+          resolve();
+        });
+      }
+
+      function onError(context: Context): void {
+        connection.removeListener("connection_error", onError);
+        debug(`Error occurred while closing amqp connection.`, context.connection.error);
+        reject(context.connection.error);
+      }
+
+      connection.once("connection_close", onClose);
+      connection.once("connection_error", onError);
+      connection.close();
+    } else {
+      resolve();
+    }
+  });
+}
+
+/**
+ * Creates an amqp session on the provided amqp connection.
+ * @param {Connection} connection The amqp connection object
+ * @return {Promise<Session>} Promise<Session>
+ * - **Resolves** the promise with the Session object when rhea emits the "session_open" event.
+ * - **Rejects** the promise with an AmqpError when rhea emits the "session_close" event while trying
+ * to create an amqp session.
+ */
+export async function createSession(connection: any): Promise<any> {
+  return new Promise((resolve, reject) => {
+    if (!connection || (connection && typeof connection !== "object")) {
+      reject(new Error("connection is a required parameter and must be of tyepe 'object'."));
```

Comment:
there's no need to put this inside of the `new Promise()` constructor. that being said, you're more than welcome to leave it the way it is. if you do, the function doesn't need to be `async`. same applies to the other functions in the file